### DISCUSSION TITLE
RegisterDelegate with runtime servicekey?

### DISCUSSION THREAD

**Discussion by zplan** [dadhi/DryIoc#549] [Q&A]
I have the following factory class which should create an object based on the service key when resolve is called:
```cs
class MyFactory
{
  MyType Create(string servicekey)
  {
  ... implementation (create myType depending on the servicekey)
  }
}
```

Is there somehow a way accessing the servicekey?
```cs
container.RegisterDelegate(typeof(MyType), r => new MyFactory().Create( r.Servicekey?? )));
```
That the consumer can call resolve like this:
```cs
var myTypeKey1 = container.Resolve<MyType>("mykey1");
var myTypeKey2 = container.Resolve<MyType>("mykey2");
```

**Comment by dadhi** [comment]
@zplan 
You can use the factory expression with `Made.Of` and pass the `ServiceKey` as an argument as following:
```cs
container.Register<MyFactory>(); // registering the factory to access it via service information below
container.Register<MyType>(Made.Of(
    r => ServiceInfo.Of<MyFactory>(), // the factory to resolve and pass as an argument to the factory method
    myFactory => myFactory.Create(Arg.Index<String>(0)), // substitute the Arg of type String with the value provided by the 0 delegate below 
    request => request.ServiceKey));
```

Check the docs here https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/SelectConstructorOrFactoryMethod.md#using-instance-factory-method

**Reply by zplan** [reply]
Hi @dadhi 
Thanks for your feedback.

Tried your proposal and it works when I resolve without a service key (but the servicekey variable is null obviously)
```cs
var myTypeKey1 = container.Resolve<MyType>();
```

But it fails when I resolve with a servicekey:
```cs
var myTypeKey1 = container.Resolve<MyType>("mykey1");
```
Exception:
```
DryIoc.ContainerException: 'code: Error.UnableToResolveFromRegisteredServices;
message: Unable to resolve resolution root MyType {ServiceKey="mykey1"}
  from container without scope
  with normal and dynamic registrations:
  (DefaultKey(0), {FactoryID=677})'
```

Any ideas?
Is it even possible to register without a ServiceKey and resolve with a ServiceKey so that the registered instance factory method is called (wich was registered without servicekey)?

**Reply by dadhi** [reply]
It is not possible to register without key but resolve with key.

**Reply by dadhi** [reply]
Looking again, you may try to use `Rules.SelectKeyedOverDefaultFactory(serviceKey)` described here: https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/RulesAndDefaultConventions.md#factoryselector
